wrappers/python/openmm/app/internal/xtc_utils/src/xtc.cpp

```diff
@@ -0,0 +1,190 @@
+/*
+MIT License
+
+Copyright (c) 2023 Accellera
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+Contributors: Stefan Doerr, Raul P. Pelaez
+*/
+#include "xtc.h"
+#include "xdrfile_xtc.h"
+#include <vector>
+#include <string>
+#include <stdexcept>
+
+// Helper functions to convert between atom/frame and x/y/z indices
+static size_t Xf(size_t atom, size_t frame, size_t nframes) {
+    return atom * 3 * nframes + frame;
+}
+
+static size_t Yf(size_t xidx, size_t nframes) {
+    return xidx + nframes;
+}
+
+static size_t Zf(size_t yidx, size_t nframes) {
+    return yidx + nframes;
+}
+
+// Helper struct to manage the XDRFILE pointer
+struct XDRFILE_RAII {
+    XDRFILE* xd;
+    XDRFILE_RAII(std::string filename, std::string mode) : xd(xdrfile_open(filename.c_str(), mode.c_str())) {
+        if (!xd) {
+            throw std::runtime_error("xtc file: Could not open file");
+        }
+    }
+
+    ~XDRFILE_RAII() {
+        if (xd)
+            xdrfile_close(xd);
+    }
+
+    operator XDRFILE*() const {
+        return xd;
+    }
+};
+
+int xtc_natoms(std::string filename) {
+    int natoms = 0;
+    if (exdrOK != read_xtc_natoms(const_cast<char*>(filename.c_str()), &natoms)) {
+        throw std::runtime_error("xtc_read(): could not get natoms\n");
+    }
+    return natoms;
+}
+
+struct XTCFrame {
+    int step;
+    float time;
+    matrix box;
+    std::vector<float> positions;
+    int natoms;
+    const float prec = 1000.0;
+    XTCFrame(int natoms) : positions(3 * natoms), natoms(natoms) {
+    }
+
+    // Read the next frame from the XTC file and store it in this object
+    int readNextFrame(XDRFILE* xd) {
+        float in_prec;
+        auto* p_ptr = reinterpret_cast<rvec*>(positions.data());
+        int status = read_xtc(xd, natoms, &step, &time, box, p_ptr, &in_prec);
+        if (prec != in_prec) {
+            throw std::runtime_error("xtc_read(): precision mismatch\n");
+        }
```

Comment:
I tracked down the test failure on Mac.  It's caused by these lines.  When you reach the end of the file, that's signaled by `read_xtc()` returning `exdrENDOFFILE`.  In that case, `in_prec` is uninitialized and doesn't match `prec`, leading it to throw an exception.

To fix it, change the check to `if (status == exdrOK && prec != in_prec)`.

Reply:
Amazing catch! would not have looked there for a long time -.- Lets see if it passes now.